File: ml_pipeline/polymorph_lamr/bench/methods.py
"""Compression methods compared by the answer-survival benchmark.

Each method takes a log chunk and a target drop rate R (fraction of tokens to
remove) and returns the compressed text. Survival is then measured on that text
(see ``survival.py``). Methods that aren't rate-tunable (deterministic dedup)
ignore R and report a single achieved ratio.

Always-available (no extra deps, GPU-free, deterministic):
  * ``DeterministicDedup``    — line-normalize + run-length collapse (a faithful
                                Python mirror of the Rust ``src/dedup.rs`` sweep).
  * ``KeepSeverityHeuristic`` — keep the most-severe lines until the token budget.
  * ``RandomDropFloor``       — deterministic pseudo-random token drop (the floor
                                any real method must beat).

Optional (skipped with a reason if the dep/model is absent):
  * ``LLMLingua2Method`` — Microsoft LLMLingua-2 token compression (needs the
                           ``llmlingua`` package; runs on CPU).
  * ``LaMRMethod``       — our trained pruner at the target rate (needs a
                           checkpoint compatible with the current single-logit
                           architecture).
"""

from __future__ import annotations

import re
import zlib
from dataclasses import dataclass
from pathlib import Path
from typing import Protocol

import tiktoken

_ENC = None


def _enc():
    global _ENC
    if _ENC is None:
        _ENC = tiktoken.get_encoding("cl100k_base")
    return _ENC


def token_count(text: str) -> int:
    return len(_enc().encode(text))


class CompressionMethod(Protocol):
    name: str
    tunable: bool  # does it respond to the target drop rate?

    def available(self) -> tuple[bool, str]:
        """(is_available, reason_if_not)."""
        ...

    def compress(self, text: str, target_drop_rate: float) -> str:
        ...


# ---------------------------------------------------------------------------
# Deterministic dedup (mirror of src/dedup.rs)
# ---------------------------------------------------------------------------

# Ordered normalization patterns — the same classes src/dedup.rs masks, so two
# lines differing only in their variable parts collapse to the same key.
_NORM_PATTERNS: list[tuple[re.Pattern, str]] = [
    (re.compile(r"\d{4}-\d{2}-\d{2}[T ]\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:?\d{2})?"), "<TS>"),
    (re.compile(r"\d{1,2}/[A-Za-z]{3}/\d{4}:\d{2}:\d{2}:\d{2}\s*[+-]\d{4}"), "<TS>"),
    (re.compile(r"\b[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}\b"), "<UUID>"),
    (re.compile(r"\b\d{1,3}(?:\.\d{1,3}){3}\b"), "<IP>"),
    (re.compile(r"\b0x[0-9a-fA-F]+\b"), "<HEX>"),
    (re.compile(r"\b\d+\b"), "<NUM>"),
]


def _normalize_key(line: str) -> str:
    key = line
    for pat, repl in _NORM_PATTERNS:
        key = pat.sub(repl, key)
    return key.strip()


@dataclass
class DeterministicDedup:
    """Collapse runs of consecutive lines sharing a normalized template into
    head + "(N lines elided)" + tail. Reversible in production (the Rust path
    caches the elided middle); here we only need the reduced text for survival.
    """

    name: str = "deterministic"
    tunable: bool = False
    min_run: int = 3  # runs longer than head+tail get an elision summary

    def available(self) -> tuple[bool, str]:
        return True, ""

    def compress(self, text: str, target_drop_rate: float) -> str:  # noqa: ARG002
        lines = text.splitlines()
        if not lines:
            return text
        out: list[str] = []
        i = 0
        n = len(lines)
        while i < n:
            key = _normalize_key(lines[i])
            j = i + 1
            while j < n and _normalize_key(lines[j]) == key:
                j += 1
            run = j - i
            if run >= self.min_run:
                out.append(lines[i])
                out.append(f"... {run - 2} lines elided ...")
                out.append(lines[j - 1])
            else:
                out.extend(lines[i:j])
            i = j
        return "\n".join(out)


# ---------------------------------------------------------------------------
# Keep-severity heuristic
# ---------------------------------------------------------------------------

_SEVERITY_ORDER = ["FATAL", "CRITICAL", "ERROR", "EXCEPTION", "TRACEBACK", "WARN"]


def _severity_rank(line: str) -> int:
    up = line.upper()
    for rank, kw in enumerate(_SEVERITY_ORDER):
        if kw in up:
            return rank
    return len(_SEVERITY_ORDER)  # non-severe


@dataclass
class KeepSeverityHeuristic:
    """Keep the most-severe lines first until the keep budget (1 - R of the input
    tokens) is spent; emit kept lines in original order. A simple, strong baseline
    for "just keep the errors."
    """

    name: str = "keep-severity"
    tunable: bool = True

    def available(self) -> tuple[bool, str]:
        return True, ""

    def compress(self, text: str, target_drop_rate: float) -> str:
        lines = text.splitlines()
        if not lines:
            return text
        total = token_count(text)
        budget = max(1, round((1.0 - target_drop_rate) * total))
        # Stable priority: (severity_rank, original_index) — severe lines first,
        # ties broken by original order.
        order = sorted(range(len(lines)), key=lambda idx: (_severity_rank(lines[idx]), idx))
        keep: set[int] = set()
        spent = 0
        for idx in order:
            cost = token_count(lines[idx]) + 1  # +1 ~ newline
            if spent + cost > budget and keep:
                continue
            keep.add(idx)
            spent += cost
            if spent >= budget:
                break
        return "\n".join(lines[i] for i in sorted(keep))


# ---------------------------------------------------------------------------
# Random-drop floor (deterministic)
# ---------------------------------------------------------------------------

def _decode_with_drop_order(
    ids: list[int],
    drop_order: list[int],
    target_drop_rate: float,
    force_keep: list[bool] | None = None,
    tokenizer: str = "cl100k",
) -> str:
    """Drop the first round(R*n) tokens in ``drop_order`` (most-droppable first),
    skipping any token marked in ``force_keep`` (the structural floor), then decode
    the survivors. Force-kept tokens are never dropped — at high R this can leave
    the achieved drop below target, which is the intended high-recall trade.
    """
    from ..label.align import decode_tokens

    n = len(ids)
    if n == 0:
        return decode_tokens(ids, tokenizer)
    k = min(n, max(0, round(max(0.0, min(1.0, target_drop_rate)) * n)))
    fk = force_keep or [False] * n
    dropped: set[int] = set()
    for i in drop_order:
        if len(dropped) >= k:
            break
        if fk[i]:
            continue
        dropped.add(i)
    return decode_tokens([tid for i, tid in enumerate(ids) if i not in dropped], tokenizer)


@dataclass
class RandomDropFloor:
    """Deterministically drop ~R of the tokens (keyed by content + index). The
    floor a real ranker must beat: same compression ratio, blind to salience.

    With ``floor=True`` the structural locker (``structural.py``) force-keeps
    high-salience atoms. A *blind* ranker plus the floor beating keep-severity is
    the non-circular demonstration that a token-level structural prior keeps
    needles that a line-level severity prior drops.
    """

    name: str = "random"
    tunable: bool = True
    floor: bool = False
    span: str | None = None        # None = token-level; else 'word' / 'chunk:N'
    aggregator: str = "max"        # empirical + Rust default (see spandecode.py)

    def __post_init__(self) -> None:
        suffix = ""
        if self.span:
            suffix += "+span"
        if self.floor:
            suffix += "+floor"
        if suffix and self.name == "random":
            self.name = "random" + suffix

    def available(self) -> tuple[bool, str]:
        return True, ""

    def compress(self, text: str, target_drop_rate: float) -> str:
        force_keep: list[bool] | None = None
        spans: list[tuple[int, int]] | None = None
        if self.floor:
            from .structural import structural_keep_mask

            ids, spans, force_keep = structural_keep_mask(text)
        elif self.span:
            from ..label.align import encode_with_spans

            ids, spans = encode_with_spans(text)
        else:
            ids = _enc().encode(text)
        if not ids:
            return text
        seed = zlib.crc32(text.encode("utf-8", "ignore"))
        # Deterministic pseudo-random per-token "drop-prob". The token-level path
        # keeps the original ascending-crc rule (most-droppable = lowest crc), so
        # mapping crc -> drop_prob must be DESCENDING in crc to match it.
        crc = [zlib.crc32(f"{seed}:{i}".encode()) & 0xFFFF_FFFF for i in range(len(ids))]
        if self.span:
            from .spandecode import span_decode

            # Higher drop_prob = more droppable. Lower crc was more droppable, so
            # invert: drop_prob = 1 - crc/2^32.
            drop_probs = [1.0 - (c / float(0x1_0000_0000)) for c in crc]
            return span_decode(
                ids, spans, text, drop_probs, target_drop_rate,
                span=self.span, aggregator=self.aggregator, force_keep=force_keep,
            )
        # Most-droppable first: ascending crc (mirrors the old threshold rule).
        order = sorted(range(len(ids)), key=lambda i: crc[i])
        return _decode_with_drop_order(ids, order, target_drop_rate, force_keep)


# ---------------------------------------------------------------------------
# LLMLingua-2 (optional)
# ---------------------------------------------------------------------------

@dataclass
class LLMLingua2Method:
    """Microsoft LLMLingua-2 token compression at the target rate. CPU-only.

    Lazily constructs the compressor on first use; ``available`` reports whether
    the package imports (the model is downloaded on first compress).
    """

    name: str = "llmlingua2"
    tunable: bool = True
    model_name: str = "microsoft/llmlingua-2-xlm-roberta-large-meetingbank"
    _compressor: object | None = None

    def available(self) -> tuple[bool, str]:
        try:
            import llmlingua  # noqa: F401
            return True, ""
        except Exception as e:  # pragma: no cover - env-dependent
            return False, f"llmlingua not importable: {e}"

    def _get(self):
        if self._compressor is None:  # pragma: no cover - heavy / env-dependent
            from llmlingua import PromptCompressor

            self._compressor = PromptCompressor(
                model_name=self.model_name, use_llmlingua2=True, device_map="cpu"
            )
        return self._compressor

    def compress(self, text: str, target_drop_rate: float) -> str:  # pragma: no cover
        rate = max(0.0, min(1.0, 1.0 - target_drop_rate))  # llmlingua "rate" = fraction kept
        comp = self._get()
        result = comp.compress_prompt(text, rate=rate, force_tokens=["\n"])
        return result.get("compressed_prompt", "") if isinstance(result, dict) else str(result)


# ---------------------------------------------------------------------------
# LaMR (our trained pruner; optional)
# ---------------------------------------------------------------------------

@dataclass
class LaMRMethod:
    """Our trained pruner: encode -> per-token drop prob -> drop the top-R by
    probability (the calibrated target-rate decode) -> decode the kept tokens.

    Needs a checkpoint compatible with the current single-logit architecture.
    """

    ckpt: Path
    name: str = "lamr"
    tunable: bool = True
    max_seq_len: int = 512  # inference window = training window (configs/modernbert.yaml)
    floor: bool = False  # apply the structural insurance floor at decode
    span: str | None = None        # None = token-level; else 'word' / 'chunk:N'
    aggregator: str = "max"        # empirical + Rust default (see spandecode.py)
    # Token space for encode/floor/decode. "auto" reads it from the checkpoint's
    # backbone (modernbert -> ModernBERT tokenizer, else cl100k); a ModernBERT
    # checkpoint MUST be fed ModernBERT ids or it sees garbage.
    tokenizer: str = "auto"
    _tokenizer: str = "cl100k"
    _model: object | None = None
    _device: object | None = None
    _err: str = ""
    # Per-text forward cache: id(text)->(ids, spans, force_keep, probs). The
    # iso-ratio gate bisects the drop rate by calling compress(same_text, varying
    # rate) ~14x; the windowed forward is identical across those calls (only the
    # decode threshold changes), so caching it turns 14 forwards into 1. Keyed by
    # the text string itself (small benchmark corpus; exact-match reuse).
    _score_cache: dict | None = None

    def __post_init__(self) -> None:
        suffix = ""
        if self.span:
            suffix += "+span"
        if self.floor:
            suffix += "+floor"
        if suffix and self.name == "lamr":
            self.name = "lamr" + suffix

    def available(self) -> tuple[bool, str]:
        if self._model is not None:
            return True, ""
        if not Path(self.ckpt).is_file():
            return False, f"checkpoint not found: {self.ckpt}"
        try:
            import torch

            from ..export.to_onnx import _load_checkpoint
            from ..train.loop import _pick_device

            model, cfg = _load_checkpoint(Path(self.ckpt))
            device = _pick_device()
            model.to(device)
            model.eval()
            self._model = model
            self._device = device
            self._torch = torch
            # Resolve the token space: "auto" follows the checkpoint's backbone.
            if self.tokenizer == "auto":
                self._tokenizer = "modernbert" if getattr(cfg, "backbone", "") == "modernbert" else "cl100k"
            else:
                self._tokenizer = self.tokenizer
            return True, ""
        except Exception as e:
            self._err = str(e)
            return False, f"load failed ({type(e).__name__}: {e})"

    def _score(self, text: str):
        """Encode ``text`` and run windowed inference to per-token drop-probs.

        Returns (ids, spans, force_keep, probs). Cached by text so the iso-ratio
        bisection (many compresses of the same text at different rates) pays the
        forward once. Windowed because real log chunks exceed the training window:
        truncating to one window would silently drop every token past it (and any
        needle in the tail — the costliest error for high recall), so the whole doc
        is scored in consecutive max_seq_len windows.
        """
        if self._score_cache is None:
            self._score_cache = {}
        cached = self._score_cache.get(text)
        if cached is not None:
            return cached
        torch = self._torch
        tok = self._tokenizer
        force_keep = None
        if self.floor:
            from .structural import structural_keep_mask

            ids, spans, force_keep = structural_keep_mask(text, tok)
        else:
            from ..label.align import encode_with_spans

            ids, spans = encode_with_spans(text, tok)
        if not ids:
            result = (ids, spans, force_keep, None)
            self._score_cache[text] = result
            return result
        import numpy as np

        W = self.max_seq_len
        chunks = []
        with torch.no_grad():
            for s in range(0, len(ids), W):
                window = ids[s : s + W]
                t_ids = torch.tensor([window], dtype=torch.long, device=self._device)
                mask = torch.ones((1, len(window)), dtype=torch.bool, device=self._device)
                chunks.append(torch.sigmoid(self._model(t_ids, mask).float())[0].cpu().numpy())
        probs = np.concatenate(chunks) if len(chunks) > 1 else chunks[0]
        result = (ids, spans, force_keep, probs)
        self._score_cache[text] = result
        return result

    def compress(self, text: str, target_drop_rate: float) -> str:
        ok, reason = self.available()
        if not ok:
            raise RuntimeError(f"LaMR unavailable: {reason}")
        tok = self._tokenizer
        ids, spans, force_keep, probs = self._score(text)
        if not ids:
            return text
        import numpy as np

        if self.span:
            from .spandecode import span_decode

            return span_decode(
                ids, spans, text, probs, target_drop_rate,
                span=self.span, aggregator=self.aggregator, force_keep=force_keep,
                tokenizer=tok,
            )
        # Calibrated target-rate decode: drop the top-k highest-prob tokens, but
        # never a force-kept structural atom.
        drop_order = np.argsort(-probs).tolist()
        return _decode_with_drop_order(ids, drop_order, target_drop_rate, force_keep, tok)


def default_methods(lamr_ckpt: Path | None = None, include_llmlingua: bool = True) -> list[CompressionMethod]:
    """The standard comparison set. LaMR is included only if a checkpoint is given;
    LLMLingua-2 only if requested (it's heavy)."""
    methods: list[CompressionMethod] = [
        DeterministicDedup(),
        KeepSeverityHeuristic(),
        RandomDropFloor(),
        RandomDropFloor(floor=True),  # structural-floor demonstration
    ]
    if include_llmlingua:
        methods.append(LLMLingua2Method())
    if lamr_ckpt is not None:
        methods.append(LaMRMethod(ckpt=Path(lamr_ckpt)))
        methods.append(LaMRMethod(ckpt=Path(lamr_ckpt), floor=True))
    return methods

File: src/methods.rs
//! Compression methods compared by the answer-survival benchmark.
//!
//! Each method takes a log chunk and a target drop rate R (fraction of tokens to
//! remove) and returns the compressed text. Survival is then measured on that
//! text (see `survival.rs`). Always-available, GPU-free, deterministic:
//!   * [`DeterministicDedup`]    — line-normalize + run-length collapse.
//!   * [`KeepSeverityHeuristic`] — keep the most-severe lines until the budget.
//!   * [`RandomDropFloor`]       — deterministic pseudo-random token drop
//!     (the floor any real ranker must beat).
//!
//! The optional torch-backed `LaMRMethod` and network `LLMLingua2Method` from the
//! Python original are intentionally out of scope for the Rust runtime. Ported
//! from `polymorph_lamr.bench.methods`.

use anyhow::Result;
use once_cell::sync::Lazy;
use regex::Regex;

use crate::spandecode::{span_decode, Aggregator};
use crate::structural::structural_keep_mask;
use crate::tokenizer::{count_tokens, decode_tokens, token_spans};

type TokenizedWithFloor = (Vec<u32>, Vec<(usize, usize)>, Option<Vec<bool>>);

/// Count cl100k tokens in `text`.
pub fn token_count(text: &str) -> usize {
    count_tokens(text).unwrap_or(0)
}

/// A compression method: encode/transform `text` to hit `target_drop_rate`.
pub trait CompressionMethod {
    fn name(&self) -> &str;
    fn tunable(&self) -> bool;
    /// `(is_available, reason_if_not)`.
    fn available(&self) -> (bool, String) {
        (true, String::new())
    }
    fn compress(&self, text: &str, target_drop_rate: f64) -> Result<String>;
}

/// Python's `round` is round-half-to-even; reproduce it for budget/k parity.
fn round_half_even(x: f64) -> i64 {
    let r = x.round();
    if (x - x.floor() - 0.5).abs() < f64::EPSILON {
        // exactly .5 — round to even
        let f = x.floor() as i64;
        if f % 2 == 0 {
            f
        } else {
            f + 1
        }
    } else {
        r as i64
    }
}

// ---------------------------------------------------------------------------
// CRC-32 (IEEE 802.3, reflected) — matches Python's `zlib.crc32`.
// ---------------------------------------------------------------------------

static CRC_TABLE: Lazy<[u32; 256]> = Lazy::new(|| {
    let mut table = [0u32; 256];
    let mut n = 0;
    while n < 256 {
        let mut c = n as u32;
        let mut k = 0;
        while k < 8 {
            c = if c & 1 != 0 {
                0xEDB8_8320 ^ (c >> 1)
            } else {
                c >> 1
            };
            k += 1;
        }
        table[n] = c;
        n += 1;
    }
    table
});

fn crc32(data: &[u8]) -> u32 {
    let mut crc = 0xFFFF_FFFFu32;
    for &b in data {
        crc = CRC_TABLE[((crc ^ b as u32) & 0xFF) as usize] ^ (crc >> 8);
    }
    crc ^ 0xFFFF_FFFF
}

// ---------------------------------------------------------------------------
// Deterministic dedup (mirror of src/dedup.rs's normalization classes)
// ---------------------------------------------------------------------------

static NORM_PATTERNS: Lazy<Vec<(Regex, &'static str)>> = Lazy::new(|| {
    vec![
        (
            Regex::new(r"\d{4}-\d{2}-\d{2}[T ]\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:?\d{2})?")
                .unwrap(),
            "<TS>",
        ),
        (
            Regex::new(r"\d{1,2}/[A-Za-z]{3}/\d{4}:\d{2}:\d{2}:\d{2}\s*[+-]\d{4}").unwrap(),
            "<TS>",
        ),
        (
            Regex::new(
                r"\b[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}\b",
            )
            .unwrap(),
            "<UUID>",
        ),
        (Regex::new(r"\b\d{1,3}(?:\.\d{1,3}){3}\b").unwrap(), "<IP>"),
        (Regex::new(r"\b0x[0-9a-fA-F]+\b").unwrap(), "<HEX>"),
        (Regex::new(r"\b\d+\b").unwrap(), "<NUM>"),
    ]
});

fn normalize_key(line: &str) -> String {
    let mut key = line.to_string();
    for (pat, repl) in NORM_PATTERNS.iter() {
        key = pat.replace_all(&key, *repl).into_owned();
    }
    key.trim().to_string()
}

/// Collapse runs of consecutive lines sharing a normalized template into
/// head + "... N lines elided ..." + tail.
pub struct DeterministicDedup {
    pub min_run: usize,
}

impl Default for DeterministicDedup {
    fn default() -> Self {
        Self { min_run: 3 }
    }
}

impl CompressionMethod for DeterministicDedup {
    fn name(&self) -> &str {
        "deterministic"
    }
    fn tunable(&self) -> bool {
        false
    }
    fn compress(&self, text: &str, _target_drop_rate: f64) -> Result<String> {
        let lines: Vec<&str> = text.lines().collect();
        if lines.is_empty() {
            return Ok(text.to_string());
        }
        let mut out: Vec<String> = Vec::new();
        let n = lines.len();
        let mut i = 0;
        while i < n {
            let key = normalize_key(lines[i]);
            let mut j = i + 1;
            while j < n && normalize_key(lines[j]) == key {
                j += 1;
            }
            let run = j - i;
            if run >= self.min_run {
                out.push(lines[i].to_string());
                out.push(format!("... {} lines elided ...", run - 2));
                out.push(lines[j - 1].to_string());
            } else {
                for ln in &lines[i..j] {
                    out.push(ln.to_string());
                }
            }
            i = j;
        }
        Ok(out.join("\n"))
    }
}

// ---------------------------------------------------------------------------
// Keep-severity heuristic
// ---------------------------------------------------------------------------

const SEVERITY_ORDER: &[&str] = &[
    "FATAL",
    "CRITICAL",
    "ERROR",
    "EXCEPTION",
    "TRACEBACK",
    "WARN",
];

fn severity_rank(line: &str) -> usize {
    let up = line.to_uppercase();
    for (rank, kw) in SEVERITY_ORDER.iter().enumerate() {
        if up.contains(kw) {
            return rank;
        }
    }
    SEVERITY_ORDER.len()
}

/// Keep the most-severe lines first until the keep budget is spent; emit kept
/// lines in original order.
#[derive(Default)]
pub struct KeepSeverityHeuristic;

impl CompressionMethod for KeepSeverityHeuristic {
    fn name(&self) -> &str {
        "keep-severity"
    }
    fn tunable(&self) -> bool {
        true
    }
    fn compress(&self, text: &str, target_drop_rate: f64) -> Result<String> {
        let lines: Vec<&str> = text.lines().collect();
        if lines.is_empty() {
            return Ok(text.to_string());
        }
        let total = token_count(text);
        let budget = round_half_even((1.0 - target_drop_rate) * total as f64).max(1);
        // Stable priority: (severity_rank, original_index).
        let mut order: Vec<usize> = (0..lines.len()).collect();
        order.sort_by_key(|&idx| (severity_rank(lines[idx]), idx));
        let mut keep: Vec<bool> = vec![false; lines.len()];
        let mut spent: i64 = 0;
        for idx in order {
            let cost = token_count(lines[idx]) as i64 + 1;
            if spent + cost > budget && keep.iter().any(|k| *k) {
                continue;
            }
            keep[idx] = true;
            spent += cost;
            if spent >= budget {
                break;
            }
        }
        let kept: Vec<&str> = lines
            .iter()
            .enumerate()
            .filter(|(i, _)| keep[*i])
            .map(|(_, l)| *l)
            .collect();
        Ok(kept.join("\n"))
    }
}

// ---------------------------------------------------------------------------
// Random-drop floor (deterministic)
// ---------------------------------------------------------------------------

/// Drop the first `round(R*n)` tokens in `drop_order` (most-droppable first),
/// skipping any token marked in `force_keep`, then decode the survivors.
fn decode_with_drop_order(
    ids: &[u32],
    drop_order: &[usize],
    target_drop_rate: f64,
    force_keep: Option<&[bool]>,
) -> Result<String> {
    let n = ids.len();
    if n == 0 {
        return decode_tokens(ids);
    }
    let rate = target_drop_rate.clamp(0.0, 1.0);
    let k = (round_half_even(rate * n as f64).max(0) as usize).min(n);
    let fk_owned;
    let fk: &[bool] = match force_keep {
        Some(f) => f,
        None => {
            fk_owned = vec![false; n];
            &fk_owned
        }
    };
    let mut dropped: std::collections::HashSet<usize> = std::collections::HashSet::new();
    for &i in drop_order {
        if dropped.len() >= k {
            break;
        }
        if fk[i] {
            continue;
        }
        dropped.insert(i);
    }
    let survivors: Vec<u32> = ids
        .iter()
        .enumerate()
        .filter(|(i, _)| !dropped.contains(i))
        .map(|(_, &tid)| tid)
        .collect();
    decode_tokens(&survivors)
}

/// Deterministically drop ~R of the tokens (keyed by content + index). With
/// `floor=true` the structural locker force-keeps high-salience atoms.
pub struct RandomDropFloor {
    name: String,
    pub floor: bool,
    pub span: Option<String>,
    pub aggregator: Aggregator,
}

impl RandomDropFloor {
    pub fn new(floor: bool, span: Option<String>, aggregator: Aggregator) -> Self {
        let mut name = String::from("random");
        if span.is_some() {
            name.push_str("+span");
        }
        if floor {
            name.push_str("+floor");
        }
        Self {
            name,
            floor,
            span,
            aggregator,
        }
    }
}

impl Default for RandomDropFloor {
    fn default() -> Self {
        Self::new(false, None, Aggregator::Max)
    }
}

impl CompressionMethod for RandomDropFloor {
    fn name(&self) -> &str {
        &self.name
    }
    fn tunable(&self) -> bool {
        true
    }
    fn compress(&self, text: &str, target_drop_rate: f64) -> Result<String> {
        let (ids, spans, force_keep): TokenizedWithFloor = if self.floor {
            let (ids, spans, fk) = structural_keep_mask(text)?;
            (ids, spans, Some(fk))
        } else if self.span.is_some() {
            let (ids, spans) = token_spans(text)?;
            (ids, spans, None)
        } else {
            let (ids, _spans) = token_spans(text)?;
            (ids, Vec::new(), None)
        };
        if ids.is_empty() {
            return Ok(text.to_string());
        }
        let seed = crc32(text.as_bytes());
        let crc: Vec<u32> = (0..ids.len())
            .map(|i| crc32(format!("{seed}:{i}").as_bytes()))
            .collect();

        if let Some(span) = &self.span {
            // Higher drop_prob = more droppable. Lower crc was more droppable, so
            // invert: drop_prob = 1 - crc/2^32.
            let drop_probs: Vec<f64> = crc
                .iter()
                .map(|&c| 1.0 - (c as f64 / 4_294_967_296.0))
                .collect();
            return span_decode(
                &ids,
                &spans,
                text,
                &drop_probs,
                target_drop_rate,
                span,
                self.aggregator,
                force_keep.as_deref(),
            );
        }
        // Most-droppable first: ascending crc.
        let mut order: Vec<usize> = (0..ids.len()).collect();
        order.sort_by_key(|&i| crc[i]);
        decode_with_drop_order(&ids, &order, target_drop_rate, force_keep.as_deref())
    }
}

/// The standard comparison set of always-available baselines.
pub fn default_methods() -> Vec<Box<dyn CompressionMethod>> {
    vec![
        Box::new(DeterministicDedup::default()),
        Box::new(KeepSeverityHeuristic),
        Box::new(RandomDropFloor::default()),
        Box::new(RandomDropFloor::new(true, None, Aggregator::Max)),
    ]
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::survival::achieved_drop_rate;

    #[test]
    fn crc32_matches_zlib_known_vectors() {
        // zlib.crc32(b"") == 0; zlib.crc32(b"123456789") == 0xCBF43926.
        assert_eq!(crc32(b""), 0);
        assert_eq!(crc32(b"123456789"), 0xCBF4_3926);
        assert_eq!(crc32(b"hello"), 0x3610_A686);
    }

    #[test]
    fn deterministic_dedup_collapses_runs_and_keeps_uniques() {
        let mut lines: Vec<String> = (0..20)
            .map(|i| format!("2023-01-01T00:00:0{}.0 INFO heartbeat ok", i % 10))
            .collect();
        lines.push("FATAL unique crash code 0xDEAD".to_string());
        let text = lines.join("\n");
        let out = DeterministicDedup::default().compress(&text, 0.0).unwrap();
        assert!(token_count(&out) < token_count(&text));
        assert!(out.contains("0xDEAD"));
        assert!(out.contains("elided"));
    }

    #[test]
    fn keep_severity_keeps_severe_line_and_shrinks() {
        let mut lines: Vec<String> = (0..40).map(|i| format!("GET /x 200 ok id={i}")).collect();
        lines.push("ERROR 500 failure token=NEEDLE42".to_string());
        let text = lines.join("\n");
        let out = KeepSeverityHeuristic.compress(&text, 0.7).unwrap();
        assert!(token_count(&out) < token_count(&text));
        assert!(out.contains("NEEDLE42"));
    }

    #[test]
    fn random_drop_is_deterministic_and_hits_rate() {
        let text = (0..400)
            .map(|i| format!("tok{i}"))
            .collect::<Vec<_>>()
            .join(" ");
        let a = RandomDropFloor::default().compress(&text, 0.5).unwrap();
        let b = RandomDropFloor::default().compress(&text, 0.5).unwrap();
        assert_eq!(a, b);
        let drop = achieved_drop_rate(&text, &a);
        assert!(0.4 < drop && drop < 0.6, "drop={drop}");
    }

    #[test]
    fn random_drop_rate_zero_keeps_everything() {
        let text = (0..100)
            .map(|i| format!("tok{i}"))
            .collect::<Vec<_>>()
            .join(" ");
        let out = RandomDropFloor::default().compress(&text, 0.0).unwrap();
        assert!(achieved_drop_rate(&text, &out) < 0.05);
    }

    #[test]
    fn floor_locks_key_anchored_value_but_not_unanchored_prose() {
        let floored = RandomDropFloor::new(true, None, Aggregator::Max);
        let mut keyed = (0..40)
            .map(|i| format!("INFO tick {i}"))
            .collect::<Vec<_>>()
            .join("\n");
        keyed.push_str("\nINFO note root_cause=\"cascading queue backpressure\" done");
        let out = floored.compress(&keyed, 0.8).unwrap();
        assert!(crate::survival::answer_survives(
            "cascading queue backpressure",
            &out
        ));

        let mut bare = (0..40)
            .map(|i| format!("INFO tick {i}"))
            .collect::<Vec<_>>()
            .join("\n");
        bare.push_str("\nthe replication follower silently fell three minutes behind schedule");
        let out2 = floored.compress(&bare, 0.8).unwrap();
        assert!(!crate::survival::answer_survives(
            "silently fell three minutes behind",
            &out2
        ));
    }
}
